exporters/autoexport/registry.go

```diff
@@ -0,0 +1,151 @@
+
+	// errUnknownExporter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExporter = errors.New("unknown exporter")
+
+	// errInvalidOTLPProtocol is returned when an invalid protocol is used in
+	// the OTEL_EXPORTER_OTLP_PROTOCOL environment variable.
+	errInvalidOTLPProtocol = errors.New("invalid OTLP protocol - should be one of ['grpc', 'http/protobuf']")
+
+	// errDuplicateRegistration is returned when an duplicate registration is detected.
+	errDuplicateRegistration = errors.New("duplicate registration")
+)
+
+// load returns tries to find the SpanExporter factory with the key and
+// then execute the factory, returning the created SpanExporter.
+// errUnknownExporter is returned if the registration is missing and the error from
+// executing the factory if not nil.
+func (r *registry) load(key string) (trace.SpanExporter, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	factory, ok := r.names[key]
+	if !ok {
+		return nil, errUnknownExporter
+	}
+	exporter, err := factory()
+	if err != nil {
+		return nil, err
+	}
+	return exporter, nil
+}
+
+// store sets the factory for a key if is not already in the registry. errDuplicateRegistration
+// is returned if the registry already contains key.
+func (r *registry) store(key string, factory func() (trace.SpanExporter, error)) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if _, ok := r.names[key]; ok {
+		return fmt.Errorf("%w: %q", errDuplicateRegistration, key)
+	}
+	r.names[key] = factory
+	return nil
+}
+
+// drop removes key from the registry if it exists, otherwise nothing.
+func (r *registry) drop(key string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	delete(r.names, key)
+}
+
+// RegisterSpanExporter sets the SpanExporter factory to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, factory func() (trace.SpanExporter, error)) {
+	if err := envRegistry.store(name, factory); err != nil {
+		// envRegistry.store will return errDuplicateRegistration if name is already
+		// registered. Panic here so the user is made aware of the duplicate
+		// registration, which could be done by malicious code trying to
+		// intercept cross-cutting concerns.
+		//
+		// Panic for all other errors as well. At this point there should not
+		// be any other errors returned from the store operation. If there
+		// are, alert the developer that adding them as soon as possible that
+		// they need to be handled here.
+		panic(err)
+	}
+}
+
+// SpanExporter returns a span exporter using the passed in name
+// from the list of registered SpanExporters. Each name must match an
+// already registered SpanExporter. A default OTLP exporter is registered
+// under both an empty string "" and "otlp".
+// An error is returned for any unknown exporters.
+func SpanExporter(name string) (trace.SpanExporter, error) {
```

Comment:
Similar here, accepting a context will allow user customization that can be passed to the exporter.